src/lib.rs

```diff
@@ -786,108 +769,85 @@ where
     phantom: std::marker::PhantomData<T>,
 }
 
-impl Buffer {
-    pub fn map_read_async<T, F>(&self, start: BufferAddress, size: BufferAddress, callback: F)
+pub struct WriteMappableBuffer {
+    // ...
+}
+
+impl WriteMappableBuffer {
+    /// Returns a writable slice of the mapped memory.
+    ///
+    /// The first time this method is called maps the memory, subsequent invocations
+    /// only return the slice.
+    ///
+    /// Note: This could be Device::map_buffer_write(&self, buffer: &mut WriteMappableBuffer) -> &[u8]
+    /// The important part is that this borrows both the mappable buffer and the device so that
+    /// the mapped memory doesn't outlive the device nor the buffer.
+    pub fn map(&mut self, device: &Device) -> &mut[u8] {
```

Comment:
I see. I suppose it is not worse than what we have today since the current callback can't access the deivce and do any submission either.
Do you think it would be possible for the callback to access the device mutably or would it be complicated/unsafe to do so?

Reply:
Feels complicated/unsafe to me

Reply:
Ok fair enough. Then I suppose passing it as non-mutable isn't worse than what we currently have, right?